third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,491 @@
+//! - removes unreachable codes
+//! (TODO: eliminate jump to branch)
+
+use move_model::model::FunctionEnv;
+use move_stackless_bytecode::{
+    function_target::FunctionData,
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{Bytecode, Label},
+    stackless_control_flow_graph::{BlockId, StacklessControlFlowGraph},
+};
+use std::collections::BTreeMap;
+
+pub struct ControlFlowGraphSimplifier {}
+
+impl FunctionTargetProcessor for ControlFlowGraphSimplifier {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv,
+        data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if fun_env.is_native() {
+            return data;
+        }
+        let mut transformer = ControlFlowGraphSimplifierTransformation::new(data);
+        transformer.transform();
+        transformer.data
+    }
+
+    fn name(&self) -> String {
+        "EliminateEmptyBlocksProcessor".to_owned()
+    }
+}
+
+struct ControlFlowGraphSimplifierTransformation {
+    data: FunctionData,
+    cfg_code_generator: ControlFlowGraphCodeGenerator,
+}
+
+impl ControlFlowGraphSimplifierTransformation {
+    pub fn new(data: FunctionData) -> Self {
+        let cfg_code_generator = ControlFlowGraphCodeGenerator::new(&data.code);
+        Self {
+            data,
+            cfg_code_generator,
+        }
+    }
+
+    /// Does the first four transformations described in the module doc
+    fn transform(&mut self) {
+        let mut elim_empty_blocks_transformer =
+            RemoveEmptyBlock::new(std::mem::take(&mut self.cfg_code_generator));
+        elim_empty_blocks_transformer.transform();
+        let mut remove_redundant_jump_transformer =
+            RemoveRedundantJump::new(elim_empty_blocks_transformer.0);
+        remove_redundant_jump_transformer.transform();
+        self.data.code = remove_redundant_jump_transformer.0.gen_code(true);
+        self.eliminate_branch_to_same_target();
+        // may introduce new empty blocks
+        let mut elim_empty_blocks_transformer =
+            RemoveEmptyBlock::new(ControlFlowGraphCodeGenerator::new(&self.data.code));
+        elim_empty_blocks_transformer.transform();
+        self.data.code = elim_empty_blocks_transformer.0.gen_code(true);
+    }
+
+    /// Transforms `if _ goto L else goto L` to `goto L`
+    fn eliminate_branch_to_same_target(&mut self) {
+        let codes = std::mem::take(&mut self.data.code);
+        self.data.code = codes
+            .into_iter()
+            .map(|bytecode| match bytecode {
+                Bytecode::Branch(attr_id, l0, l1, _) if l0 == l1 => Bytecode::Jump(attr_id, l0),
+                _ => bytecode,
+            })
+            .collect();
+    }
+}
+
+#[derive(Default)]
+struct ControlFlowGraphCodeGenerator {
+    cfg: StacklessControlFlowGraph,
+    code_blocks: BTreeMap<BlockId, Vec<Bytecode>>,
```

Comment:
Please document the fields. Also, is `code_blocks` not redundant with information present in `cfg`?

Reply:
`cfg` doesn't own the code, so we have to own another `code: Vec<Bytecode>`. But `code: BTreeMap<BlockId, Vec<Bytecode>>` is easier to manipulate, as we don't just transform the code linearly from front to end, say we have a empty block to remove, we need to transform its predecessor blocks, which maybe code block already visited.

Also I should doc that `BlockContent` is invalidated during the transformation.

Reply:
done